src/lightning/pytorch/cli.py

```diff
@@ -228,6 +228,17 @@ def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> Non
         if self.already_saved:
             return
 
+        self.save_config(trainer, pl_module, stage)
+        if trainer.is_global_zero:
+            # save only on rank zero to avoid race conditions.
+            self.extra_save_config(trainer, pl_module, stage)
+            self.already_saved = True
+
+        # broadcast so that all ranks are in sync on future calls to .setup()
+        self.already_saved = trainer.strategy.broadcast(self.already_saved)
+
+    def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+        """Saves the config to the log_dir only on rank zero."""
```

Comment:
Why this refactor instead of just adding `self.extra_save_config()` after `self.parser.save()` in the original implementation?

Reply:
Could be done differently. This is just my first idea given broadcasts in the code which I don't fully know how to deal with yet. Two main objectives are:
- Allow users to implement a custom save without having to worry about broadcast.
- Have the choice to keep or not the standard save to `log_dir`.

Another idea I had was with a single method, say `save_config`. It would be called from `setup` inside the `if trainer.is_global_zero:` block. If someone subclasses `SaveConfigCallback` and implements `save_config`, they would need to call `super().save_config` to preserve the standard save to `log_dir`. The problem is that the standard save broadcasts `log_dir` and `file_exists`, which means that it can't be called from inside the `if trainer.is_global_zero:`. We would need a standard save that does not require any broadcast.

Any ideas?

Reply:
For customizing the `log_dir`, I would suggest a simple `log_dir: Optional[str]` argument in the callback init

Then in setup it does:

```python
if self.log_dir is None:
    log_dir = trainer.log_dir  # this broadcast the directory
else:
    log_dir = broadcast(log_dir)
```

I think the `save_config` idea is good, perhaps with an `on_after_save_config` hook that replaces your `save_config_extra` idea


```python
def save_config(self, config_path):
    self.parser.save(
        self.config, config_path, skip_none=False, overwrite=self.overwrite, multifile=self.multifile
    )


...

        # save the file on rank 0
        if trainer.is_global_zero:
            fs.makedirs(log_dir, exist_ok=True)
            self.save_config(config_path)
            self.on_after_save_config(config_path)
            self.already_saved = True
```

Reply:
The signature for `save_config` needs to be

```python
def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
```

Otherwise it wouldn't be possible extend to save in a logger or for other use cases people might come up with.

To disable the standard save, an init parameter could be, but like  `save_to_log_dir: bool = True`. Though it isn't an ideal solution. It is less simple to disable. In the subclass you would need to implement `__init__` to hard code `super().__init__(*args, save_to_log_dir=False, **kwargs)` or provide this when instantiating the LightningCLI class.

Reply:
I pushed a new implementation for this.